### PR TITLE
Add the idea of a jar manager to handle jar_dependencies

### PR DESCRIPTION
**Pull request by purbon**
In #2937 and #2954 we discussed the idea to build a jar manager in order to change the way we're loading the jar dependencies within our code base.

This PR add the concept of a jar manager responsible of finding and loading jars as dependencies.

Collaterally this PR also removed the load_elasticsearch_jars! method as it's a legacy code that was there from the early days when the ES jars were shipped with LS and not in the plugins. 

This new jar manager is in use in

* https://github.com/logstash-plugins/logstash-input-log4j/pull/8
* https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/101
* https://github.com/logstash-plugins/logstash-output-elasticsearch-ec2/pull/6
* https://github.com/logstash-plugins/logstash-input-drupal_dblog/pull/7

Relates to #2974

For now it still need a bit more test to be sure everything is working as expected.

### CODE REVIEW

**Comment by purbon** [comment]
This test scenario has been successful for me.

* Install all plugins
* Update the plugins that has been changed in this PR.
* Run the all plugins test.

:green_apple:

File: lib/logstash/environment.rb [hunk 1]
```diff
@@ -45,7 +45,6 @@ module Environment
 
     LOGSTASH_HOME = ::File.expand_path(::File.join(::File.dirname(__FILE__), "..", ".."))
     JAR_DIR = ::File.join(LOGSTASH_HOME, "vendor", "jar")
```
Comment (wiibaa) [inline lib/logstash/environment.rb]: I suppose you could also remove this line, JAR_DIR is not used anymore in 1.5

Comment (purbon) [inline lib/logstash/environment.rb]: makes sense @wiibaa

**Comment by colinsurprenant** [comment]
- why not keep `load_jars!` in the `Envronment` module? this is pretty much the idea of this module...
- why default to a new `dependencies/jar` path? we already have the `vendor` dir for that? just `vendor/jars` no?

**Comment by purbon** [comment]
* I wanted to create something separate in order to let this be loaded without the overhead of the hole LS-Environment, I don't specifically see this as LS env, while loading JARS we don't need stuff like bundle for example. But this was just my point or initial idea. 

* I wanted to made this generic enough  so you can change this if you want, this are just the default parameters I thought interesting. But the important question here is, do we want the jars to live within the lib dir? or within the vendors? I don't have strong opinions, but I always use them throw lib and don't strongly dislike the idea.

**Comment by colinsurprenant** [comment]
Agreed about bundler code not belonging in the `Environment` module and PR #2752 solves that.

On the other hand since the `Environment` module is already loaded in all plugins I don't think we need to created a separate module for a single jar loading method which we basically already had with the `Environment,load_elasticsearch_jars!` method. Also, I this it is specifically about the environment: it is basically just a path convention and checking that we are running JRuby. 

For the jars dir, first, let's make the default relevant to the logstash environment. Once we agree and where we should store the jars, let's change the defaults to that. For the `vendor/` dir, I am actually on the fence about it, since typically you expect the be able to delete the `vendor` dir and recreate it using some rake tasks or something. On the other hand, this is exactly what the `vendor` dir is about, to store external dependencies.  So I think I'd vote for putting jars into `vendor/jar` unless we see a potential problem with that?

**Comment by purbon** [comment]
* About ``Environment``` ok, I like the idea to make it there.
* About the directory, I see the point and the regular usage of the ```vendor``` dir, as you said I'm on the fence, but find the idea interesting ... so lets use vendor too. Is this something we all like?

**Comment by jsvd** [comment]
+1 for vendor/jars

**Comment by colinsurprenant** [comment]
@jsvd did you mean `vendor/jar` ?

**Comment by purbon** [comment]
why not ```vendor/jar-dependencies/runtime``` ?

**Comment by jsvd** [comment]
yes. yes I did.

**Comment by colinsurprenant** [comment]
@purbon `vendor/jar-dependencies/runtime` wat?

**Comment by purbon** [comment]
what happen when we have also other dependencies for test? for example.

**Comment by purbon** [comment]
All related code in here has been adapted to the last review comments, including the dependants PR's on the plugins side.

**Comment by jsvd** [comment]
image

**Comment by purbon** [comment]
@jsvd I agree I should add them, but before I wanted us to agree on the API, although it seems there is no concern with it.

**Comment by purbon** [comment]
@jsvd simple unit test suite added.

File: lib/logstash/environment.rb [hunk 2]
```diff
@@ -114,6 +95,42 @@ def bundler_setup!(options = {})
       ::Bundler.setup
     end
 
+    def runtime_jars_root(dir_name, package)
```
Comment (jsvd) [inline lib/logstash/environment.rb]: In the plugins we do:

```
ROOT_DIR   = File.expand_path(File.join(File.dirname(__FILE__), ".."))
LogStash::Environment.load_runtime_jars! File.join(ROOT_DIR, "vendor")
```

But that's a lot of unecessary code we need on each plugin. At most, we need `LogStash::Environment.load_runtime_jars! File.dirname(__FILE__)` and even that we could somehow find by ourselves.

If we reduce this to just `LogStash::Environment.load_runtime_jars!`, we no longer need this extra file and can just include the require call in the plugin code. thoughts?

Comment (purbon) [inline lib/logstash/environment.rb]: I agree with you in part, however I did it this way in order to provide more flexibility, keep in mind now this file has a ```ROOT_DIR``` just one level below the file that load the jars, but it can happen to be in another location. Or that for some jars, we aim to store them in another place. 

In my opinion, even if It might make sense, I find it too restrictive.

File: lib/logstash/environment.rb [hunk 3]
```diff
@@ -114,6 +95,42 @@ def bundler_setup!(options = {})
...
+      ::File.join(dir_name, package, "runtime-jars")
+    end
+
+    def test_jars_root(dir_name, package)
+      ::File.join(dir_name, package, "test-jars")
+    end
+
+    def load_runtime_jars!(dir_name="vendor", package="jar-dependencies")
```
Comment (jsvd) [inline lib/logstash/environment.rb]: the code of `load_runtime_jars!` and `load_test_jars!` is the same, it should be put in a common method, leaving them as 

```ruby
def load_runtime_jars!(dir_name="vendor", package="jar-dependencies")
  generic_load_jars!(runtime_jars_root(dir_name, package))
end

def load_test_jars!(dir_name="vendor", package="jar-dependencies")
  generic_load_jars!(test_jars_root(dir_name, package))
end
```

File: lib/logstash/environment.rb (hunk 2 above)
Comment (jsvd) [inline lib/logstash/environment.rb]: I understand the flexibility argument, but don't really see a practical use case for it. The jars should be in a predetermined folder, and all .jar files we find in that folder and its subfolders get loaded.

Comment (purbon) [inline lib/logstash/environment.rb]: Ok, lets do it that way so we can move on and merge, we are always allowed to revert the change if need.

**Comment by wiibaa** [comment]
@purbon there is an empty spec file `spec/util/environment_spec.rb`, maybe this PR could clean it up ?

**Comment by purbon** [comment]
good catch again @wiibaa, cleaning this up too.

File: lib/logstash/environment.rb (hunk 2 above)
Comment (jsvd) [inline lib/logstash/environment.rb]: After discussing with @purbon, let's go with the current strategy. The amount of code plugins need to load their jars should either be a one liner or it could be detected automatically. However this requires a different strategy, so I've opened an issue to deal with this: https://github.com/elastic/logstash/issues/3005

Comment (purbon) [inline lib/logstash/environment.rb]: Good @jsvd, this is a good discussion :-)

**Comment by jsvd** [comment]
Reviewed, ran specs, tested with logstash-output-elasticsearch ([pr/101](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/101)) using :path and `bin/plugin install --no-verify`

LGTM

File: lib/logstash/environment.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 require "logstash/errors"
-require 'logstash/version'
```
Comment (ph) [inline lib/logstash/environment.rb]: double *dragon* quotes +1

File: lib/logstash/environment.rb [hunk 5]
```diff
@@ -114,6 +95,43 @@ def bundler_setup!(options = {})
...
+    def test_jars_root(dir_name, package)
+      ::File.join(dir_name, package, "test-jars")
+    end
+
+    def load_runtime_jars!(dir_name="vendor", package="jar-dependencies")
+      load_jars!("#{runtime_jars_root(dir_name, package)}/*.jar")
+    end
+
+    def load_test_jars!(dir_name="vendor", package="jar-dependencies")
+      load_jars!("#{test_jars_root(dir_name, package)}/*.jar")
```
Comment (ph) [inline lib/logstash/environment.rb]: `/*.jar` this is unix specific no?

Comment (purbon) [inline lib/logstash/environment.rb]: good point @ph, changing that to use a non platform specific constant

File: lib/logstash/environment.rb [hunk 6]
```diff
@@ -114,6 +95,43 @@ def bundler_setup!(options = {})
...
+      jar_files = find_jars(pattern)
+      require_jars! jar_files
+    end
+
+    def find_jars(pattern)
+      require 'java'
+      jar_files = Dir.glob(pattern)
+      raise(LogStash::EnvironmentError, "Could not find jar files under #{pattern}") if jar_files.empty?
+      jar_files
+    end
```
Comment (ph) [inline lib/logstash/environment.rb]: The actual messages should be defined in the locale yml.

File: spec/core/environment_spec.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    it "find test jars in the default location" do
+      expect(subject).to receive(:find_jars).with(default_test_location).and_return([])
+      subject.load_test_jars!
+    end
+
+    context "when loading a jar file" do
+
+      let(:dummy_jar_file)          { "vendor/jar-dependencies/runtime-jars/elasticsearch.jar" }
```
Comment (ph) [inline spec/core/environment_spec.rb]: many spaces?

Comment (purbon) [inline spec/core/environment_spec.rb]: probably :-), stupid editors

**Comment by ph** [comment]
So if I understand correctly this code, this will load all the jars in the vendor directory?
Even if we don't actually use the plugin?

**Comment by purbon** [comment]
@ph it load the plugins whenever this functions are used, in the current case from a file within each plugins, so no, it will load the jars whenever the plugins are used for first time.

**Comment by purbon** [comment]
@ph can you make sure the code I took from here related to maven makes sense?

File: lib/logstash/environment.rb (hunk 5 above)
Comment (ph) [inline lib/logstash/environment.rb]: No need for constant, `File.join` will take care of that.

```
irb(main):001:0> File.join('vendor', '*.jar')
=> "vendor/*.jar"

Comment (purbon) [inline lib/logstash/environment.rb]: ok

Comment (ph) [inline lib/logstash/environment.rb]: `File::SEPARATOR` is the constant for the platform specific stuff, but `File.join` doesnt care if the path exist or not.

**Comment by ph** [comment]
Well, in my current checkout this PR breaks `logstash-*-kafka`

```
~/e/logstash git:master ❯❯❯ bin/plugin install logstash-input-kafka                                                                                                                                                                                                                                                                                                   ⬆ ✱
Validating logstash-input-kafka
Installing logstash-input-kafka
Installation successful
~/e/logstash git:master ❯❯❯ tree vendor/bundle/jruby/1.9/gems/logstash-input-kafka-0.1.12                                                                                                                                                                                                                                                                             ⬆ ✱
vendor/bundle/jruby/1.9/gems/logstash-input-kafka-0.1.12
├── CONTRIBUTORS
├── DEVELOPER.md
├── Gemfile
├── LICENSE
├── README.md
├── Rakefile
├── lib
│   └── logstash
│       └── inputs
│           └── kafka.rb
├── logstash-input-kafka.gemspec
└── spec
    └── inputs
        └── kafka_spec.rb

5 directories, 9 files
```

**Comment by purbon** [comment]
@ph good you have test with kafka now, might be is because i removed the maven code, because I tested before that and it was working ... see thanks for the review man

**Comment by purbon** [comment]
@ph I checked with the change and with master and I see the same number of files for the kafka input, I see no change in this plugin ... I must say there was actually no change with this plugin for this PR

**Comment by ph** [comment]
Did a test with this PR, Jars are in jruby-kafka.

LGTM.

**Comment by purbon** [comment]
ok, merging this in :shipit:  :shipit:  :shipit:  :shipit:  :shipit:  :shipit:

**Comment by purbon** [comment]
test this please

**Comment by purbon** [comment]
can you test this please

**Comment by codekitchen** [comment]
Related to this change, it looks like the documentation describing how to add jar dependencies to a logstash plugin is now out of date: https://www.elastic.co/guide/en/logstash/current/_how_to_write_a_logstash_input_plugin.html#_jar_dependencies

Is there guidance on the new correct way of doing things? I'm trying to update my plugin https://github.com/codekitchen/logstash-input-kinesis to work with logstash 1.5.1, but I'm not clear on how to get my jars properly vendored now. Thanks.

**Comment by purbon** [comment]
@codekitchen you are right, documentation should be updated. Do you mind opening an issue for that? 

On the other side, you can see how to add you own jars if you take a look for example at https://github.com/logstash-plugins/logstash-input-log4j/tree/master/lib , there you will see a final named ```logstash-input-log4j_jars.rb``` who has the code necessary to load the jars.

Jars are then placed into the gem, in the case of the ```logstash-input-log4j``` you can find them at https://github.com/logstash-plugins/logstash-input-log4j/tree/master/vendor/jar-dependencies/runtime-jars . You can apply the same pattern in your plugin.

Makes sense? 

/cheers

**Comment by codekitchen** [comment]
Thanks @purbon I filed https://github.com/elastic/logstash/issues/3455

I very rarely work with the JVM, so it actually took me a little while to figure out a good method for getting the jars I need into `vendor` -- I found it a lot more difficult than the previous `jar-dependencies` based method where everything just seemed to work based on a gemspec declaration. I also found no scripts or clues in the other plugins such as elasticsearch or log4j, which surprised me a bit. How were those jars vendored?

Eventually I ended up getting it working by adding a `pom.xml` file to my project and using `mvn` to vendor the jars, in https://github.com/codekitchen/logstash-input-kinesis/commit/816a10671ae44b3342580263f43f378511bb5322 , but I have no idea if that's the best option.

**Comment by purbon** [comment]
@codekitchen good you figure it out, basically nowadays we just drop all need jars in a folder, even if not the best solution, ```jar-dependencies``` was causing us some random hassle during plugin management and installation. We're working with the maintainer of the gem to address this issues sooner or later and go back to it as it's clear a better way to handle jar dependencies.

In the meanwhile this is a necessary pain point to handle plugins who need jars. In a nutshell you should grab all jars you need and put them in the runtime folder. 

Thanks a lot for filling #3455, this would help our documentation efforts.
